DataFormats/GEMRecHit/src/GEMCSCSegment.cc

```diff
@@ -0,0 +1,124 @@
+GEMCSCSegment::GEMCSCSegment(const CSCSegment* csc_segment, const std::vector<const GEMRecHit*> gem_rhs, LocalPoint origin, LocalVector direction, AlgebraicSymMatrix errors, double chi2) : 
+
+  RecSegment(buildDetId(csc_segment->cscDetId())),
+  theOrigin(origin), 
+  theLocalDirection(direction), theCovMatrix(errors), theChi2(chi2) {
+
+  for(unsigned int i=0; i<gem_rhs.size(); ++i) {
+    theGEMRecHits.push_back((*gem_rhs[i]));
+  }
+  theCSCSegment = *csc_segment;
+  // LogDebug
+  edm::LogVerbatim("GEMCSCSegment")<< "[GEMCSCSegment :: ctor] CSCDetId: " << csc_segment->cscDetId() << " CSC RecHits: " <<csc_segment->specificRecHits().size()
+				   << " GEM RecHits: " << gem_rhs.size()<<"\n" //  << " Fit chi2: "<<chi2<<" Position: "<<origin<<" Direction: "<<direction
+				   << "    CSC Segment Details: \n"<<*csc_segment<<"\n"
+				   << " GEMCSC Segment Details: \n"<<*this<<"\n"
+				   << "[GEMCSCSegment :: ctor] ------------------------------------------------------------";
+}
+
+
+GEMCSCSegment::~GEMCSCSegment() {}
+
+
+std::vector<const TrackingRecHit*> GEMCSCSegment::recHits() const{
+
+  std::vector<const TrackingRecHit*> pointersOfRecHits;
+  for (std::vector<GEMRecHit>::const_iterator irh = theGEMRecHits.begin(); irh!=theGEMRecHits.end(); ++irh) {
+    pointersOfRecHits.push_back(&(*irh));
+  }
+  for (std::vector<CSCRecHit2D>::const_iterator irh = theCSCSegment.specificRecHits().begin(); irh!=theCSCSegment.specificRecHits().end(); ++irh) {
+    pointersOfRecHits.push_back(&(*irh));
+  }
+  return pointersOfRecHits;
+}
+
+std::vector<TrackingRecHit*> GEMCSCSegment::recHits() {
+
+  std::vector<TrackingRecHit*> pointersOfRecHits;
+  for (std::vector<GEMRecHit>::iterator irh = theGEMRecHits.begin(); irh!=theGEMRecHits.end(); ++irh) {
+    pointersOfRecHits.push_back(&(*irh));
+  }
+  /*
+  for (std::vector<CSCRecHit2D>::const_iterator irh = theCSCSegment.specificRecHits().begin(); irh!=theCSCSegment.specificRecHits().end(); ++irh) {
+    pointersOfRecHits.push_back(&(*irh));
+    }*/
+  return pointersOfRecHits;
+}
+
+
+LocalError GEMCSCSegment::localPositionError() const {
+  return LocalError(theCovMatrix[2][2], theCovMatrix[2][3], theCovMatrix[3][3]);
+}
+
+
+LocalError GEMCSCSegment::localDirectionError() const {
+  return LocalError(theCovMatrix[0][0], theCovMatrix[0][1], theCovMatrix[1][1]); 
+}
+
+
+AlgebraicVector GEMCSCSegment::parameters() const {
+  // For consistency with DT, CSC and what we require for the TrackingRecHit interface,
+  // the order of the parameters in the returned vector should be (dx/dz, dy/dz, x, z)
+  
+  AlgebraicVector result(4);
+
+  result[0] = theLocalDirection.x()/theLocalDirection.z();
+  result[1] = theLocalDirection.y()/theLocalDirection.z();    
+  result[2] = theOrigin.x();
+  result[3] = theOrigin.y();
+
+  return result;
+}
+
+
+AlgebraicMatrix GEMCSCSegment::projectionMatrix() const {
+  static AlgebraicMatrix theProjectionMatrix( 4, 5, 0);
+  static bool isInitialized = false;
```

Comment:
Actually, this is a thread safety problem. Since isInitialized is not declared std::atomic the compiler is allowed to reorder the instructions such that 'isInitialized' is set to true before the values of theProjectionMatrix are actually changed.